### PR TITLE
Add Support for Cast Function on MV Columns

### PR DESCRIPTION
**Pull request by SabrinaZhaozyf**
Label: `feature`
This PR adds support for Cast transform function for mv columns.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java [hunk 1]
```diff
@@ -84,6 +84,22 @@ public void init(List<TransformFunction> arguments, Map<String, DataSource> data
         case "JSON":
           _resultMetadata = JSON_SV_NO_DICTIONARY_METADATA;
           break;
+        case "INTEGERMV":
+        case "INTMV":
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:88]: We should probably not need to introduce this. Won't look clean in the SQL query.

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:88]: We should be able to detect that column being casted is SV or MV. `INTMV` is no type and looks like a hack

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java [hunk 2]
```diff
@@ -30,6 +33,70 @@
...
+    String[][] expectedStringValues = new String[NUM_ROWS][];
+    ArrayCopyUtils.copy(innerLayerInt, expectedStringValues, NUM_ROWS);
+    testCastTransformFunctionMV(transformFunction, expectedStringValues);
+
+    expression = RequestContextUtils.getExpression(String.format("arrayMax(cAst(%s AS INT))", DOUBLE_MV_COLUMN));
+    transformFunction = TransformFunctionFactory.get(expression, _dataSourceMap);
+    FieldSpec.DataType resultDataType = transformFunction.getResultMetadata().getDataType();
+    Assert.assertEquals(resultDataType, FieldSpec.DataType.INT);
+
+    expression = RequestContextUtils.getExpression(String.format("arraySum(cAst(%s AS INT))", DOUBLE_MV_COLUMN));
```
Comment (SabrinaZhaozyf) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java:89]: TODO: add comment

**Comment by siddharthteotia** [comment]
Yet to review the main logic. Thanks for addressing the main concern of INTMV v/s INT

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java (hunk 1 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java]: Done.

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java (hunk 2 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java:89]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java [hunk 3]
```diff
@@ -47,40 +49,51 @@ public void init(List<TransformFunction> arguments, Map<String, DataSource> data
...
+          _resultMetadata = isSVCol ? FLOAT_SV_NO_DICTIONARY_METADATA : FLOAT_MV_NO_DICTIONARY_METADATA;
           break;
         case "DOUBLE":
-          _resultMetadata = DOUBLE_SV_NO_DICTIONARY_METADATA;
+          _resultMetadata = isSVCol ? DOUBLE_SV_NO_DICTIONARY_METADATA : DOUBLE_MV_NO_DICTIONARY_METADATA;
           break;
         case "DECIMAL":
         case "BIGDECIMAL":
         case "BIG_DECIMAL":
+          // TODO: MV cast to BIG_DECIMAL type
```
Comment (somandal) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:73]: just curious, why not throw an exception here like you do for other types which don't have support for MV such as BOOLEAN/TIMESTAMP?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java [hunk 4]
```diff
@@ -96,6 +109,204 @@ public TransformResultMetadata getResultMetadata() {
...
+          _stringValuesMV = new String[length][];
+        }
+        if (inputDataType == DataType.BOOLEAN) {
+          int[][] intValues = _transformFunction.transformToIntValuesMV(projectionBlock);
+          for (int i = 0; i < length; i++) {
+            int rowLen = intValues[i].length;
+            for (int j = 0; j < rowLen; j++) {
+              _stringValuesMV[i][j] = Boolean.toString(intValues[i][j] == 1);
+            }
+          }
```
Comment (somandal) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:165]: nit: I know you only call this once, but might still be good to put this into a separate private method for easier code readability. same for DataType.TIMESTAMP handled next.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java (hunk 3 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java]: Good catch! Will throw an exception as well because we decided to address this TODO in a separate PR in the future after our discussion.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java (hunk 4 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java]: Done - simplified the code for both SV and MV.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java [hunk 5]
```diff
@@ -47,40 +47,54 @@ public void init(List<TransformFunction> arguments, Map<String, DataSource> data
...
         case "DOUBLE":
-          _resultMetadata = DOUBLE_SV_NO_DICTIONARY_METADATA;
+          _resultMetadata = isSVCol ? DOUBLE_SV_NO_DICTIONARY_METADATA : DOUBLE_MV_NO_DICTIONARY_METADATA;
           break;
         case "DECIMAL":
         case "BIGDECIMAL":
         case "BIG_DECIMAL":
+          if (!isSVCol) {
+            // TODO: MV cast to BIG_DECIMAL type
+            throw new IllegalArgumentException("MV cast is not supported for type - " + targetType);
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:73]: (nit) suggest changing the message to `"Cast is not supported on multi-value column to target type: " + targetType`

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:73]: Same elsewhere where you are throwing exception

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java [hunk 6]
```diff
@@ -30,6 +33,71 @@
...
+    transformFunction = TransformFunctionFactory.get(expression, _dataSourceMap);
+    Assert.assertTrue(transformFunction instanceof CastTransformFunction);
+    assertEquals(transformFunction.getName(), CastTransformFunction.FUNCTION_NAME);
+    int[][] innerLayerInt = new int[NUM_ROWS][];
+    ArrayCopyUtils.copy(_floatMVValues, innerLayerInt, NUM_ROWS);
+    float[][] expectedFloatValues = new float[NUM_ROWS][];
+    ArrayCopyUtils.copy(innerLayerInt, expectedFloatValues, NUM_ROWS);
+    testCastTransformFunctionMV(transformFunction, expectedFloatValues);
+
+    expression = RequestContextUtils.getExpression(
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java:71]: Thanks for adding the nested cast transform tests. 

Can you please also add another nested cast transform where the outer transform is not a CAST but some other transform ? For example, for the internal use case we want to use it inside `ID_SET(CAST(mv_string_col AS LONG))`. So, a similar test would be great.

File: pinot-core/src/test/java/org/apache/pinot/queries/CastQueriesTest.java [hunk 7]
```diff
@@ -130,6 +146,18 @@ public void testCastSum() {
     assertEquals(((Number) aggregationResult.get(1)).intValue(), NUM_RECORDS / 4);
   }
 
+  @Test
+  public void testCastMV() {
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/CastQueriesTest.java:150]: Add an `ID_SET` related test here to ensure the functionality we want to support for the internal use case works.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/ArrayCopyUtils.java [hunk 8]
```diff
@@ -239,4 +278,259 @@ public static void copy(byte[][] src, BigDecimal[] dest, int length) {
       dest[i] = BigDecimalUtils.deserialize(src[i]);
     }
   }
+  public static void copy(int[][] src, int[][] dest, int length) {
```
Comment (siddharthteotia) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/ArrayCopyUtils.java:281]: Is it possible to do this refactoring in a separate PR unless it is absolutely needed ?

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java (hunk 6 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java:71]: Since ID_SET is an aggregation function, I will add the test under the corresponding aggregation function test.

Comment (SabrinaZhaozyf) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CastTransformFunctionTest.java:71]: Done

File: pinot-core/src/test/java/org/apache/pinot/queries/CastQueriesTest.java (hunk 7 above)
Comment (SabrinaZhaozyf) [reply pinot-core/src/test/java/org/apache/pinot/queries/CastQueriesTest.java:153]: Done.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/ArrayCopyUtils.java (hunk 8 above)
Comment (SabrinaZhaozyf) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/ArrayCopyUtils.java:281]: Done - reverted refactoring in BaseTransformFunction. Will put up a separate PR.

**Review by siddharthteotia — APPROVED** [review]
lgtm
